File: training/src/slopmarker/data/chunking.py
"""Deterministic chunking of a paragraph block into scoreable units (scope.md 7.3).

This is what the extension does at inference time, so it is a coupling point pinned by
fixtures/windows.json. Keep it separate in your head from the *training* window sampler
(data/windows.py), which is stochastic and never ported.
"""

from __future__ import annotations

from .normalize import collapse_whitespace, count_words
from .sentences import split_sentences

MIN_WORDS = 40
MAX_WORDS = 400


def chunk_block(text: str, min_words: int = MIN_WORDS, max_words: int = MAX_WORDS) -> list[str]:
    """Split one block into <= max_words chunks at sentence boundaries.

    Blocks under `min_words` are skipped entirely (return []). A single sentence longer
    than `max_words` is emitted alone rather than cut mid-sentence -- that is the only
    way a chunk may exceed the cap.

    A trailing chunk may come out under `min_words`. That is deliberate: merging it back
    into the previous chunk would breach `max_words`, and scope.md 8 already shrinks
    short chunks toward 0.5 by `words / 80`, so a short tail can only extend a run, never
    start one.
    """
    text = collapse_whitespace(text)
    if count_words(text) < min_words:
        return []

    chunks: list[str] = []
    current: list[str] = []
    current_words = 0
    for start, end in split_sentences(text):
        sentence = text[start:end]
        words = count_words(sentence)
        if current and current_words + words > max_words:
            chunks.append(" ".join(current))
            current, current_words = [], 0
        current.append(sentence)
        current_words += words
    if current:
        chunks.append(" ".join(current))
    return chunks

File: extension/src/shared/chunking.ts
/**
 * Deterministic chunking of a paragraph block into scoreable units (scope.md 7.3).
 *
 * Port of training/src/slopmarker/data/chunking.py, pinned by fixtures/windows.json
 * (`chunk_cases`). Keep it separate in your head from the *training* window sampler
 * (data/windows.py), which is stochastic and is never ported.
 */

import { collapseWhitespace, countWords, toCodePoints } from "./normalize.js";
import { splitSentences } from "./sentences.js";

export const MIN_WORDS = 40;
export const MAX_WORDS = 400;

/**
 * Split one block into <= maxWords chunks at sentence boundaries.
 *
 * Blocks under `minWords` are skipped entirely (return []). A single sentence longer than
 * `maxWords` is emitted alone rather than cut mid-sentence -- that is the only way a chunk
 * may exceed the cap.
 *
 * A trailing chunk may come out under `minWords`. That is deliberate: merging it back into
 * the previous chunk would breach `maxWords`, and scope.md 8 already shrinks short chunks
 * toward 0.5 by `words / 80`, so a short tail can only extend a run, never start one.
 */
export function chunkBlock(
  text: string,
  minWords: number = MIN_WORDS,
  maxWords: number = MAX_WORDS,
): string[] {
  const collapsed = collapseWhitespace(text);
  const cp = toCodePoints(collapsed);
  return chunkSpans(collapsed, minWords, maxWords).map(([a, b]) => cp.slice(a, b).join(""));
}

/**
 * The same chunking, as [start, end) codepoint spans into `collapseWhitespace(text)`.
 *
 * The content script needs this: to highlight a run it has to map chunks back to DOM
 * ranges, and it cannot do that from the strings alone. Spans are safe because
 * `splitSentences` partitions the input and its sentences rejoin with exactly one space
 * (`" ".join(sentences) === text.trim()`), so a group of consecutive sentences is always a
 * contiguous span rather than something that merely stringifies the same.
 *
 * chunking.test.ts asserts the two agree on every fixture case, so this cannot drift into
 * being a second, subtly different chunker.
 */
export function chunkSpans(
  collapsed: string,
  minWords: number = MIN_WORDS,
  maxWords: number = MAX_WORDS,
): Array<[number, number]> {
  if (countWords(collapsed) < minWords) return [];

  const cp = toCodePoints(collapsed);
  const spans: Array<[number, number]> = [];
  let start: number | null = null;
  let end = 0;
  let currentWords = 0;

  for (const [sentenceStart, sentenceEnd] of splitSentences(collapsed)) {
    const words = countWords(cp.slice(sentenceStart, sentenceEnd).join(""));
    if (start !== null && currentWords + words > maxWords) {
      spans.push([start, end]);
      start = null;
      currentWords = 0;
    }
    start ??= sentenceStart;
    end = sentenceEnd;
    currentWords += words;
  }
  if (start !== null) spans.push([start, end]);
  return spans;
}
